src/vga.js

```diff
@@ -740,115 +953,467 @@ VGAScreen.prototype.complete_redraw = function()
+
+        switch(this.crtc_mode & 0x3)
+        {
+            case 0x2:
+                // Alternating rows using bit 13
+                // Assumes max scan line = 1
+                row = row << 1 | (addr >> 13 & 0x1);
+                break;
+            case 0x1:
+                // Alternating rows using bit 14
+                // Assumes max scan line = 3
+                row = row << 1 | (addr >> 14 & 0x1);
+                break;
+            case 0x0:
+                // Cycling through rows using bit 13 and 14
+                // Assumes max scan line = 3
+                row = row << 2 | (addr >> 13 & 0x3);
+                break;
+        }
+
+        // Reassemble address
+        return row * this.virtual_width + col + (this.start_address << shift_count);
+    }
+    else
+    {
+        // Convert to 1 pixel per address
+        return addr << shift_count;
+    }
+};
+
+VGAScreen.prototype.scan_line_to_screen_row = function(scan_line)
+{
+    // Double scanning. The clock to the row scan counter is halved
+    // so it is not affected by the memory address bit substitutions below
+    if(this.max_scan_line & 0x80)
+    {
+        scan_line >>>= 1;
+    }
+
+    // Maximum scan line, aka scan lines per character row
+    // This is the number of repeats - 1 for graphic modes
+    var repeat_factor = 1 + (this.max_scan_line & 0x1F);
+    scan_line = Math.ceil(scan_line / repeat_factor);
+
+    // Odd and Even Row Scan Counter
+    // Despite repeated address counter values, because bit 13 of the shifted
+    // address is substituted with bit 0 of the row scan counter, a different
+    // display buffer address is generated instead of repeated
+    // Assumes maximum scan line register is set to 2 or 4.
+    // Note: can't assert this as register values may not be fully programmed.
+    if(!(this.crtc_mode & 0x1))
+    {
+        scan_line <<= 1;
+    }
+
+    // Undo effects of substituted bit 14
+    // Assumes maximum scan line register is set to 2 or 4
+    // Note: can't assert this as register values may not be fully programmed.
+    // Other maximum scan line register values would result in weird addressing
+    // anyway
+    if(!(this.crtc_mode & 0x2))
+    {
+        scan_line <<= 1;
+    }
+
+    return scan_line;
+};
+
+/**
+ * @param {number} cols_count
+ * @param {number} rows_count
+ */
+VGAScreen.prototype.set_size_text = function(cols_count, rows_count)
+{
+    this.max_cols = cols_count;
+    this.max_rows = rows_count;
+
+    this.bus.send("screen-set-size-text", [cols_count, rows_count]);
+};
+
+VGAScreen.prototype.set_size_graphical = function(width, height, bpp, virtual_width, virtual_height)
+{
+    this.screen_width = width;
+    this.screen_height = height;
+
+    this.stats.bpp = bpp;
+    this.stats.is_graphical = true;
+    this.stats.res_x = width;
+    this.stats.res_y = height;
+
+    this.bus.send("screen-set-size-graphical", [width, height, bpp, virtual_width, virtual_height]);
```

Comment:
bpp is ignored by the screen adapter, could you remove it?

Reply:
It appears that `bpp` is listened for here:

https://github.com/copy/v86/blob/6440f7f9bcbf04535af8e3d8d0044a411d5640f9/src/browser/main.js#L972-L976

Should we keep `bpp` or still remove it from the bus message?